framework/src/main/java/org/tron/core/db/BandwidthProcessor.java

```diff
@@ -24,10 +25,12 @@
 public class BandwidthProcessor extends ResourceProcessor {
 
   private Manager dbManager;
+  private ChainBaseManager chainBaseManager;
 
-  public BandwidthProcessor(Manager manager) {
+  public BandwidthProcessor(Manager manager, ChainBaseManager chainBaseManager) {
```

Comment:
the variable dbManager of BandwidthProcessor should be removed completely. The implementation that both Manager and ChainBaseManager reserved used as the combination inheritance is redundancy. So I advise that implementing the function getHeadSlot() in the ChainbaseManager and then removing the Manager from the  BandwidthProcessor construction function.

Reply:
> getHeadSlot

Your suggestion is right. But the BandwidthProcessor still has some dependency with Manager. Please refer to genesisBlock. Another PR is needed.

Reply:
yes, I know. So, I think you should determine the role of the Chainbase at first. If it is the container of the XXXStore, I think there can be a delegate function using the genesisBlock as an input parameter to implement the getHeadSlot().

Reply:
Currently, you can't implement getHeadSlot in ChainBaseManager, it has dependency on args. The genesisBlock should also be moved to ChainBaseManager. And the logic is quite complexed. So I plan to implement it in the next PR.

Reply:
Ok, let's make a review deeply and determine whether there can be another reasonable way.

Reply:
> Ok, let's make a review deeply and determine whether there can be another reasonable way.

Waiting for your quick reply. Thanks.

Reply:
> the variable dbManager of BandwidthProcessor should be removed completely. The implementation that both Manager and ChainBaseManager reserved used as the combination inheritance is redundancy. So I advise that implementing the function getHeadSlot() in the ChainbaseManager and then removing the Manager from the BandwidthProcessor construction function.

I have removed dbManager from BandwidthProcessor completely. Please check.